### PR TITLE
API,MAINT: Make ``NpyIter_GetTransferFlags`` public and avoid old uses

### PR DESCRIPTION
**Pull request by seberg**
This combines two related things:
1. Make `NpyIter_GetTransferFlags` a public function.  Now that we have these flags in the public API, there seems no reason not to make this public.
2. The iterator currently found the `NEEDS_API` flag every time (with `REFS_OK`) for use in `NpyIter_IterationNeedsAPI`.  But we should generally _not_ use `NpyIter_IterationNeedsAPI` anymore in NumPy.  The reason is that we need to know what buffering needs, but already know anyway if we need the API for whatever we do with the dtypes.  So:
   * Change code to not use it or add a comment for why it is used.
   * Remove the flag and move the logic to `NpyIter_IterationNeedsAPI` so that we avoid the unnecessary flag discovery (at least in throttled CPU mode ~5% of a trivial reduce).

I completely removed one macro, because I didn't want the confusion of which version is used (or two macros).

**This is ready for review, but marking as draft, because the reduce code move is probably only safe after gh-27883.  (if we are lucky, the sanitizer CI run notices this)**

### CODE REVIEW

**Review by mattip — APPROVED** [review]
LGTM, once it is rebased and tests pass.

**Comment by seberg** [comment]
Ping @ngoldbaum for awareness though, since it adds the public function.

Also, will push one small change: The return value should be `NPY_ARRAYMETHOD_FLAGS` rather than int.
(After that will merge, the circleci/benchmark failures are due to a new `scipy-doctest` and the lint also has a general failure right now.)

**Comment by seberg** [comment]
OK, put this in very happy to follow up if the docs should be improved (or anything else).

**Pull request merged**
